Implement CALL-IN-MAIN-THREAD

lokedhs opened the pull request:
This function allows a multi-threaded module to perform actions on the main
event handler thread and should be used whenever a different thread
needs to invoke any stumpwm functions.

Currently, only SBCL is supported.

PuercoPop commented:
To provide further context I'll paste the conversation on `#stumpwm`

<details>
  <summary> Click to Expand </summary>
```
15:23 <loke> [03:06:56] Is there a way to force the repaint of the status
             bar?
15:23 <mood> [03:50:38] loke: Perhaps (update-all-mode-lines) ?
15:23 <loke> [03:51:27] mood: Can I call that from a different thread?
15:23 <loke> [03:51:42] In general, how much of stump is thread-safe?
15:23 <mood> [04:35:51] loke: I wouldn't assume anything is thread-safe, to
             be honest
15:23 <loke> [04:37:06] mood: I thought so.
15:23 <loke> [04:37:24] Is there a method I can use to queue a function to
             be called in the event handler thread?
15:23 <mood> [04:37:36] I'm not sure if CLX is even thread-safe
15:23 <loke> [04:38:28] mood: That's the point. So I'd need a
             CALL-IN_MAIN-THREAD function or something.
15:23 <mood> [04:39:06] loke: You could probably work something out with the
             io loop, though I'm not sure how hard it would be
15:23 <mood> [04:39:39] Yet another area to work on. I should make myself a
             real todo list for this
15:23 <loke> [04:39:52] mood: If something like that doesn't exist, I'd be
             happy to see if I can implement it.
15:23 <loke> [04:40:03] It's definitely needed in my dbus module.
15:23 <loke> [04:40:14] Since there are lots of asynchrnous events
             happening.

15:44 <PuercoPop> loke: the io-loop has a 'job queue', the entry point is
                  RUN-WITH-TIMER but the API is not very good. (Maybe we
                  should copy sb-timer?). Any feedback would be
                  appreciated. https://github.com/stumpwm/stumpwm/blob/master/stumpwm.lisp#L90
22:43 <loke> PuercoPop: I'm looking at it, and it's not very reliable.
22:43 <loke> I create a timer for 1 second and it sometimes takes 5 seconds
             for it to trigger
22:43 <loke> (or more)
22:44 <loke> Also, RUN-WITH-TIMER isn't thread-safe in itself.
22:45 <loke> Ah, I see why that happens. It's because the io-channel timeout
             is only updated when the main thread does something. Makes
             sense.
22:47 <loke> I would suggest creating a pipe, and then an io channel
             listeners on that pipe. Jobs would be scheduled by adding
             closures to a list, and then send a message on the pipe,
             indicating that there are waiting jobs. The main thread can
             then drain that list by funcalling the closures.
23:03 [PuercoPop nods]
23:06 <PuercoPop> So a 'job channel'
23:13 <loke> PuercoPop: Yes. Exactly.
23:14 <loke> The job queue management would be the only thing that would
             need proper synchronisation. Everything else can be called
             though the queue.
23:27 <loke> PuercoPop: I started working on an implementation of this idea.
23:27 <loke> How, I need to open a pipe. The question is, I can do it using
             SBCL-specific code, but there is no generic way of doing it.
23:28 <loke> Adding a dependency on iolib would not be an option, I guess.
23:31 <PuercoPop> loke: ideally the iolib code should be pluggable so we
                  could have a contrib to use iolib
23:31 <PuercoPop> I've been meaning to suggest using cl-launch so we can use
                  quicklisp so we can clean-up some code.
23:32 <PuercoPop> But having iolib as a dependency in the main repo would be
                  a no go
23:33 <loke> Well, the job queue would be a central part of the
             application. After all, it needs to be the main way to run
             things, so it can't be an optional thing.
23:33 <loke> This means, that if the solution will be built on pipes, a way
             to create that pipe needsw to be found.
23:34 <loke> One altnerative would be to use platform-specific code, but I
             only know how to do it in SBCL
23:34 <loke> (which is SB-POSIX:PIPE)
```
</details>

jorams commented:
I'll have to look more closely at this later, but some quick initial ideas:

- We might not want bordeaux-threads as a dependency right now.
- CCL has `ccl::pipe`, which does what we want but sadly isn't exported.
- xcvb uses `linux:pipe` for this on CLISP, but no `linux` package exists in my CLISP.
- This appears to currently break StumpWM on all implementations other than SBCL (because of the `open-pipe` call in `stumpwm-internal-loop`), which we don't want.
- Maybe on implementations that don't support this functionality, `call-in-main-thread` could use `run-with-timer`.

lokedhs commented:
The most recent commit addresses the concerns raised above. While it's not actually been tested, it should work on CCL now. There is also a fallback which is no worse than what we had before. It calls the functions directory rather than using the timer. The timer should be a bit more safe, but if used the function will not invoked until the next event, which could take some time. Should the timer still be used?

dbjergaard commented:
Hi,

We can't add dependencies to stumpwm (but paulownia is a different story).  Also, we can't add something that only works in sbcl, and has untested solutions in ccl and no solution in ecl and gcl.  

I know this is a frustrating development constraint (which is why paulownia will only support sbcl) but a number of users run stumpwm from esoteric lisp flavors, and so doing non-generic things is usually a non-starter.

So, the bar for getting merged is "must not break the build on ccl and sbcl" which this passes, but in reality the bar is a little higher because I don't want to break ecl or gcl compatability.

jorams commented:
I did some testing:

- CCL currently crashes during startup because `io-channel-ioport` can't handle CCL file descriptors

    ```
    > Error: There is no applicable method for the generic function:
    >          #<STANDARD-GENERIC-FUNCTION IO-CHANNEL-IOPORT #x18F9347E>
    >        when called with arguments:
    >          (#<STUMPWM::XLIB-IO-LOOP #x19312816> #<FD-BINARY-INPUT-STREAM :ISO-8859-1 (PIPE/5) #x19311FEE>)
    > While executing: #<CCL::STANDARD-KERNEL-METHOD NO-APPLICABLE-METHOD (T)>, in process listener(1).
    > Type :GO to continue, :POP to abort, :R for a list of available restarts.
    > If continued: Try calling it again
    > Type :? for other options.
    1 > :b
     (F6FFED70) : 0 (FUNCALL #'#<#<CCL::STANDARD-KERNEL-METHOD NO-APPLICABLE-METHOD (T)>> #<STANDARD-GENERIC-FUNCTION IO-CHANNEL-IOPORT #x18F9347E> #<STUMPWM::XLIB-IO-LOOP #x19312816> #<FD-BINARY-INPUT-STREAM :ISO-8859-1 (PIPE/5) #x19311FEE>) 135
     (F6FFED98) : 1 (FUNCALL #'#<#<STANDARD-METHOD IO-CHANNEL-IOPORT (T STUMPWM::REQUEST-CHANNEL)>> #<STUMPWM::XLIB-IO-LOOP #x19312816> #<STUMPWM::REQUEST-CHANNEL #x193115DE>) 79
     (F6FFEDAC) : 2 (FUNCALL #'#<#<STANDARD-METHOD IO-LOOP-ADD (STUMPWM::XLIB-IO-LOOP T)>> #<STUMPWM::XLIB-IO-LOOP #x19312816> #<STUMPWM::REQUEST-CHANNEL #x193115DE>) 39
     (F6FFEDBC) : 3 (STUMPWM-INTERNAL-LOOP) 647
     (F6FFEDE8) : 4 (STUMPWM-INTERNAL ":1") 3119
     (F6FFEE3C) : 5 (STUMPWM ":1") 383
     (F6FFEE4C) : 6 (CALL-CHECK-REGS STUMPWM) 247
     (F6FFEE68) : 7 (CHEAP-EVAL (STUMPWM)) 87
     (F6FFEE84) : 8 (FUNCALL #'#<(:INTERNAL CCL::EVAL-STRING CCL::STARTUP-CCL)> "(stumpwm:stumpwm)") 431
     (F6FFEEA8) : 9 (STARTUP-CCL ("home:ccl-init" "home:\\.ccl-init")) 1431
     (F6FFEED8) : 10 (FUNCALL #'#<(:INTERNAL (CCL:TOPLEVEL-FUNCTION (CCL::LISP-DEVELOPMENT-SYSTEM T)))>) 55
     (F6FFEEE8) : 11 (FUNCALL #'#<(:INTERNAL CCL::MAKE-MCL-LISTENER-PROCESS)>) 551
     (F6FFEF34) : 12 (RUN-PROCESS-INITIAL-FORM #<TTY-LISTENER listener(1) [Active] #x18203636> (#<CCL:COMPILED-LEXICAL-CLOSURE # #x1820336E>)) 639
     (F6FFEF78) : 13 (FUNCALL #'#<(:INTERNAL (CCL::%PROCESS-PRESET-INTERNAL (CCL:PROCESS)))> #<TTY-LISTENER listener(1) [Active] #x18203636> (#<CCL:COMPILED-LEXICAL-CLOSURE # #x1820336E>)) 519
     (F6FFEFCC) : 14 (FUNCALL #'#<(:INTERNAL CCL::THREAD-MAKE-STARTUP-FUNCTION)>) 255
    1 >   
    ```

- CLISP starts up fine, but any command requiring an input window errors with `NIL isn't a INT16`. That error also occurs on `master` though, so this PR is not at fault. I opened issue #287 for this. Note that I had to rebase this code onto `master` to even get there, because of #284.

I think using timers on "unsupported" implementations would be better. Just calling the function and completely killing thread safety doesn't seem sane to me, because the entire purpose of using `call-in-main-thread` is that you want some level of safety. Timers can provide a little more safety, and functions like this by definition can't really guarantee *when* your code will get execute, so it seems to be the better option.

lokedhs commented:
Fair enough. I'll make the appropriate changes to address these two points.

lokedhs commented:
The timer workaround is now used on platforms that don't have pipes support implemented. I also moved the platform-specific code to `wrappers.lisp` where this code actually seems to belong.

dbjergaard commented:
Is this ready to merge?

lokedhs commented:
It should be. I'm using it already and it is stable on SBCL.

dbjergaard commented on wrappers.lisp:
```diff
@@ -383,4 +383,56 @@ regarding files in sysfs. Data is read in chunks of BLOCKSIZE bytes."
 
+(defun open-pipe (&key (element-type '(unsigned-byte 8)))
+  "Create a pipe and return two streams. The first value is the input
+stream, and the second valus is the output stream."
```
valus -> value

jorams commented on stumpwm.lisp:
```diff
@@ -219,6 +273,16 @@ The action is to call FUNCTION with arguments ARGS."
+
+         ;; If we have no implementation for the current CL, then
+         ;; don't register the channel.
+         #+sbcl
```
This should be `#+(or sbcl ccl)` right?

jorams replied:
The answer is *no*. I forgot about the fact that the non-dummy IO loop is currently SBCL only.

jorams commented:
I tried running with this on SBCL, CCL, ECL and CLISP. It does not (appear to) introduce any new issues, so I'm merging.
